torch_np/_dtypes.py

```diff
@@ -0,0 +1,174 @@
+        else:
+            raise TypeError(f"data type '{name}' not understood")
+        self._name = _name
+
+    @property
+    def name(self):
+        return self._name
+
+    def __eq__(self, other):
+        if isinstance(other, dtype):
+            return self._name == other.name
+        else:
+            other_instance = dtype(other)
+            return self._name == other_instance.name
+
+    def __repr__(self):
+        return f'dtype("{self.name}")'
+
+    __str__ = __repr__
+
+
+dt_names = ['float16', 'float32', 'float64',
+         'complex64', 'complex128',
+         'uint8',
+         'int8',
+         'int16',
+         'int32',
+         'int64',
+         'bool']
+
+
+dt_aliases_dict = {
+    'i1' : 'int8',
+    'i2' : 'int16',
+    'i4' : 'int32',
+    'i8' : 'int64',
+    'b'  : 'int8',   # XXX: srsly?
+    'f2' : 'float16',
+    'f4' : 'float32',
+    'f8' : 'float64',
+    'c8' : 'complex64',
+    'c16': 'complex128',
+}
+
+
+python_types_dict = {
+    int: 'int64',
+    float: 'float64',
+    builtins.bool: 'bool'
+}
+
+
+float16 = dtype("float16")
+float32 = dtype("float32")
+float64 = dtype("float64")
+complex64 = dtype("complex64")
+complex128 = dtype("complex128")
+uint8 = dtype("uint8")
+int8 = dtype("int8")
+int16 = dtype("int16")
+int32 = dtype("int32")
+int64 = dtype("int64")
+bool = dtype("bool")
+
+
+# Map the torch-suppored subset dtypes to local analogs
+# "quantized" types not available in numpy, skip
+_dtype_from_torch_dict = {
+        # floating-point
+        torch.float16: 'float16',
+        torch.float32: 'float32',
+        torch.float64 : 'float64',
+        # np.complex32 does not exist
+        torch.complex64: 'complex64',
+        torch.complex128: 'complex128',
+        # integer, unsigned (unit8 only, torch.uint32 etc do not exist)
+        torch.uint8: 'uint8',
+        # integer
+        torch.int8: 'int8',
+        torch.int16: 'int16',
+        torch.int32: 'int32',
+        torch.int64: 'int64',
+        # boolean
+        torch.bool : 'bool'
+}
+
+
+# reverse mapping
+_torch_dtype_from_dtype_dict = {_dtype_from_torch_dict[key]: key
+                                for key in _dtype_from_torch_dict}
+
+
+def dtype_from_torch(torch_dtype):
+    try:
+        name = _dtype_from_torch_dict[torch_dtype]
+        return dtype(name)
+    except KeyError:
+        # mimic numpy: >>> np.dtype('unknown') -->  TypeError
+        raise TypeError
+
+
+def torch_dtype_from(dtyp):
+    if dtyp is None:
+        return None
+    if isinstance(dtyp, dtype):
+        name = dtyp.name
+    else:
+        name = dtyp
+    try:
+        return _torch_dtype_from_dtype_dict[name]
+    except KeyError:
+        # mimic numpy: >>> np.dtype('unknown') -->  TypeError
+        raise TypeError
+
+
+
+# The casting below is defined *with dtypes only*, so no value-based casting!
+
+# These two dicts are autogenerated with autogen/gen_dtypes.py,
+# using numpy version 1.23.5.
+
+_can_cast_dict = {
```

Comment:
Implementation detail: no structured casting between categories, no differentiating between arrays and scalars (scalar can upcast arrays).